src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+
+  // variables to create before alg
+  int[][] timeMatrix;
+  HashMap<Integer, String> intToPlaceId;
+  HashMap<Integer, OpeningHours.Period> openHours;
+  int intOfWeek; // Sunday is 0, Saturday is 6
+
+  // var to output
+  Tuple finalAnswer;
+ 
+  /**
+   * Initializes context and intOfWeek, must call solver then getFinalAnswer(). 
+   * 
+   * @param context geoContext from TripServlet
+   * @param intOfWeek day of week as an int (Sunday:0 - Saturday:6) from TripDay
+   */
+  public TspSolver(GeoApiContext context, int intOfWeek) {
+    if (context == null) {
+      throw new IllegalArgumentException("GeoApiContext is null");
+    } else if (intOfWeek < 0 || intOfWeek > 6) {
+      throw new IllegalArgumentException("intOfWeek not in valid range (0-6)");
+    }
+
+    // set class constants from TripServlet
+    this.context = context;
+    this.intOfWeek = intOfWeek;
+  }
+
+  /**
+   * Solves the problem, initializes various variables needed to solve, sets finalAnswer
+   * 
+   * @param center central location
+   * @param pois list of pois to visit
+   */
+  public void solver(String center, List<String> pois) throws IOException {
+    if (pois == null) {
+      throw new IllegalArgumentException("pois input list is null");
+    } else if (pois.isEmpty()) {
+      throw new IllegalArgumentException("pois input list is empty");
+    } else if (center == null) {
+      throw new IllegalArgumentException("center poi is null");
+    } else if (center.isEmpty()) {
+      throw new IllegalArgumentException("center poi is empty string");
+    }
+
+    // variables used for tracking
+    int currPos = START_POS;
+    int numNodes = pois.size() + 1; // total nodes: #locations + center
+    LocalTime currentTime = START_TIME;
+    int count = START_COUNT;
+    Tuple cost = new Tuple(START_COST, new ArrayList<>());
+    Tuple ans = new Tuple(START_ANS, new ArrayList<>());
+    boolean[] visited = new boolean[numNodes]; // sets values to "false"  
+
+    // call function to set timeMatrix, intToPlaceId, & openHours
+    createTimeMatrix(center, pois);
+    populateIntMapAndOpenHours(center, pois);
+
+    // set central visited node to visted
+    visited[0] = true;
+
+    // call recursive helper fuction to get final Answer
+    this.finalAnswer = solverHelper(currPos, numNodes, currentTime, count, 
+        cost, ans, visited);
+  }
+
+  /**
+   * Recursive helper function that is used to solve computational part of
+   * algorithm. Called in global `solver()` method
+   *
+   * @param currPos represents the placeId of where the alg currently is
+   * @param numNodes total number of locations to visit (includes center)
+   * @param currentTime time if path at this point were taken 
+   * @param count number of pois visited during this path
+   * @param cost Tuple of current path (time cost, List of pois)
+   * @param ans the current best solution found
+   * @param array of which pois have been visited by this path
```

Comment:
optional: these descriptions are pretty clear. But I think it would still be easier to understand this method if it had fewer arguments. I think you can actually skip some of these arguments without too much trouble.

Eg: currPos will always be equal to the final element in `cost.getCurrPath()`. Similarly count is equal to `cost.getCurrPath().size()` and I think numNodes will always be equal to `visited.length`.

Reply:
The issue with this is that the first iteration has size=0 and then currPos is trying to get index 0 from an empty list

Reply:
Fixed by adding 0 (index of central location) into the list at the start of the recursion